File: 30.substring-with-concatenation-of-all-words.py
class Solution:
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        ret = []
        wl = len(words[0])
        wc = len(words)
        if len(s) < wl * wc:
            return ret
        occurs = {}
        for w in words:
            if w in occurs:
                occurs[w] += 1
            else:
                occurs[w] = 1
        for i in range(len(s) - wl * wc + 1):
            occ = {}
            cnt = 0
            for j in range(wc):
                w = s[i + j * wl:i + j * wl + wl]
                if not w in occurs:
                    break
                else:
                    if w in occ:
                        occ[w] += 1
                    else:
                        occ[w] = 1
                    if occ[w] == occurs[w]:
                        cnt += 1
                    elif occ[w] > occurs[w]:
                        break
            if cnt == len(occurs):
                ret.append(i)
        return ret

File: 30.substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> ret;
        int wl = words[0].length();
        int wc = words.size();
        if (s.length() < wl * wc) {
            return ret;
        }
        map<string, int> occurs;
        for (auto& w : words) {
            occurs[w]++;
        }
        for (int i = 0; i <= s.length() - wl * wc; i++) {
            map<string, int> occ;
            int cnt = 0;
            for (int j = 0; j < wc; j++) {
                string w = s.substr(i + j * wl, wl);
                if (occurs.find(w) == occurs.end()) {
                    break;
                } else {
                    occ[w]++;
                    if (occ[w] == occurs[w]) {
                        cnt++;
                    } else if (occ[w] > occurs[w]) {
                        break;
                    }
                }
            }
            if (cnt == occurs.size()) {
                ret.push_back(i);
            }
        }
        return ret;
    }
};